Development for multioutput regression

charlesfu4 opened the pull request:
Multioutput regression based on #292 revision according to what mfeurer mentioned. look like there are still some part missing, I will try to figure it out.

charlesfu4 commented:
just figured out there are more parts need to be revised

- metrics.py
- smbo.py
and so on

charlesfu4 commented:
Long time no see. I have revised some part of the pipelines that caused crashes last time. Now it works on a multioutput regression of a 96-output problem for me. 

- Note that the **ensemble_memory_limit** should be adjusted larger. 
     For example, fitting trian_X, train_y with size (35605, 112), (35605, 96), and **ensemble_nbest** can be only set to 12 with ensemble memory limit given 2048MB. 

- There are minor issues happening during fitting process
    For example sometimes it got stuck at the final time when the SMAC providing **challenger** not better than the previous one. Not sure what caused this.

The warning messages are shown below:
[WARNING] [2020-05-02 00:38:08,675:smac.intensification.intensification.Intensifier] Challenger was the same as the current incumbent; Skipping challenger
[WARNING] [2020-05-02 00:38:08,675:smac.intensification.intensification.Intensifier] Challenger was the same as the current incumbent; Skipping challenger

I would like to add xgboost regressor back and see what's the difference next time.

charlesfu4 commented:
I found out my revision works on the 0.6.0 version but not the development version. 
In the development version, it would only output Dummy regressor after the same running time. I think that's because the new budget functionality that caused this problem. I will figure out how to fix it.

charlesfu4 commented:
**Problem:** The version did not synchronize with remote automl upstream.
**Causes:** Caused by my unfamiliarity of git development procedure, I never fetch remote upstream every time I have revision in local and push it directly to my origin development.
**Solution:** after multiple rebase this version is at 0.7.0 and not behind the remote automl upstream

**Fruther contribution needed:**
1. Mulitoutput regression only picks regreesors natively support multioutput, not sure show add Multioutput wrapper function of sklearn for the others or not, should try it 
2. Feature preprocessing has not yet implemented multioutput part, but in my test locally it still somehow works. This should be dealt later on.

charlesfu4 commented:
Hi @mfeurer, I will appreciate if there's any advice for this pull request, thanks!

mfeurer commented on autosklearn/automl.py:
```diff
@@ -926,14 +922,18 @@ def _check_X(self, X):
+
+        if y.ndim == 1:
+            return y
         if y.ndim == 2 and y.shape[1] == 1:
```
Please use elif here

mfeurer commented on autosklearn/automl.py:
```diff
@@ -926,14 +922,18 @@ def _check_X(self, X):
             y = np.ravel(y)
+            return y
 
+        self._n_outputs = y.shape[1]
```
Could you please briefly describe why this is necessary here?

charlesfu4 replied:
I think this is redundant. Did not find it out. Thanks!

charlesfu4 replied:
Just found out this should be modified and moved to **AutoMLRegressor** part. Under the circumstance when using cross validation, the refit function check if self._n_output matches targets' _n_output. 
self._n_output was only initialized as 1 and I did not update it there in Regressor function, now solved.

mfeurer commented on autosklearn/automl.py:
```diff
@@ -1089,41 +1087,44 @@ def predict_proba(self, X, batch_size=None, n_jobs=1):
-            metric = r2
+        y_task = type_of_target(y)
+        task = self._task_mapping.get(y_task)
+        print(task)
```
Could you please drop this print statement?

mfeurer commented on autosklearn/metrics/__init__.py:
```diff
@@ -57,6 +60,8 @@ def __call__(self, y_true, y_pred, sample_weight=None):
             y_pred[y_pred > 0.5] = 1.0
             y_pred[y_pred <= 0.5] = 0.0
+        elif type_true == 'continuous-multioutput':
+            y_pred = y_pred
```
Could you please put a pass here instead?

charlesfu4 commented:
Thanks for the advices and corrections!
> I guess it would be good to check this actually now. A simple way to do so is to add a new unit test in test/test_pipeline/test_regression.py following the pattern of test_configurations (or test_multiclass in the respective classifition test). That'll randomly sample configurations and fail if they are invalid.

Sure I will do that. Will taking **dataset.load_linenerud** as the multioutput regression testing data good enough? Or a random generated multioutput regression target will be better?  


> Also, it would probably be good if you test the AutoSklearnRegressor with multilabel output in test/test_automl/test_estimators to ensure that it will continue to work in the future.

Do you mean **multilabel-indicator**? From my understanding, shouldn't multioutput regression include **continuous-multioutput** only or at most include also **multiclass-multioutput**?

charlesfu4 commented:
- Added label **handles_multioutput**:  
For all the models under **feature_preprocessing**, **data_preprocessing**, and **classification**(all set to False). Whether each fp or dp model supports multioutput was considered by sklearn documentation.
   
- Added **multioutput unit test** and fixed bugs:

  - Find out there were few bugs: Under **pipeline/regression.py**, includes a repeated function get_available_components, which also appears in **pipeline/components/regression/__init__.py**.

  - Under **pipeline/components/regression/__init__.py**, a redundant term data_prop was replaced as dataset_properties.
  - Add 'handels_multioutput' in ThirdPartComponents should_be_there of **pipeline/components/base.py**
  - Add 'handels_multioutput' in DummyClassifier and DummyPreprocessor in **test/test_pipeline/test_classification.py**
  - Add _test_multioutput_ in **test/test_pipeline/test_regression.py**, with random generated data (20 features and 4 targets)

- Problem with **Kernel_PCA**: 
I tried to dodge the error several times by changing the train_size_maximum. But it's better to solve it in the scikit-learn back end.
scikit-learn/scikit-learn#16718 solved it and I think it is only available until scikit-learn=0.23, test unit test almost always failed I I think this could be solved after making this version compatible to scikit-learn=0.23

charlesfu4 commented:
Hi @mfeurer would you review my code? Thanks!
Also, I found out that if I fed resampling strategy with sklearn BaseCrossValidator like KFold or TimeSeriesSplit object, it always returned Dummy classifier and Dummy regressor. I also tested them on 0.7.0 master version without my multi-output regression modification. The results were the same. I think I can try to look into it and fix it, should I open an issue for it? Thank you!

mfeurer commented on autosklearn/pipeline/components/data_preprocessing/minority_coalescense/minority_coalescer.py:
```diff
@@ -38,6 +38,7 @@ def get_properties(dataset_properties=None):
                 'handles_classification': True,
                 'handles_multiclass': True,
                 'handles_multilabel': True,
+                'handles_multioutput': False,
```
Hey, could you please let me know why this doesn't work for multi-output regression?

mfeurer commented on autosklearn/pipeline/components/feature_preprocessing/truncatedSVD.py:
```diff
@@ -39,6 +39,7 @@ def get_properties(dataset_properties=None):
                 'handles_classification': True,
                 'handles_multiclass': True,
                 'handles_multilabel': True,
+                'handles_multioutput': False,
```
Hey, truncatedSVD shouldn't look at the y values, so it should be able to handle multi-output regression.

charlesfu4 replied:
Yeah, I think I kind of missed these two, I've updated them. Thanks!

charlesfu4 commented:
> Hey @charlesfu4 this looks really great, thanks for your work!
> 
> I have to minor questions and then we're good to merge this I believe. We'll do a release today (0.7.1) and will then include your changes in the next release (0.7.2).
> 
> Regarding your comments:
> 
> * yes, we're aware of the kernelPCA problem and unfortunately it'll only be resolved once we upgrade scikit-learn to 0.23
> * yes, it would be great if you could have a look into why it fails with the KFold resampling strategy. I don't think you need to look into the time series split as that one is not really working well yet anyway.

It's my pleasure to participate in the development, thank you for the code reviews and advices.
